ffi/src/lib.rs

```diff
@@ -709,6 +757,334 @@ pub extern "C" fn visit_expression_literal_long(
+    asbox!(scan as boxed_scan => {
+        let kernel_state = boxed_scan.kernel_scan.global_scan_state();
+        BoxHandle::into_handle(GlobalScanState { kernel_state })
+    })
+}
+
+/// # Safety
+///
+/// Caller is responsible for passing a valid handle.
+#[no_mangle]
+pub unsafe extern "C" fn drop_global_scan_state(state: *mut GlobalScanState) {
+    unsafe {
+        drop(Box::from_raw(state));
+    }
+}
+
+// Intentionally opaque to the engine.
+#[allow(clippy::type_complexity)]
+pub struct KernelScanDataIterator {
+    // Box -> Wrap its unsized content this struct is fixed-size with thin pointers.
+    // Item = Box<dyn EngineData>, see above, Vec<bool> -> can become a KernelBoolSlice
+    data: Box<dyn Iterator<Item = DeltaResult<(Box<dyn EngineData>, Vec<bool>)>>>,
+
+    // Also keep a reference to the external client for its error allocator.
+    // Parquet and Json handlers don't hold any reference to the tokio reactor, so the iterator
+    // terminates early if the last table client goes out of scope.
+    engine_interface: Arc<dyn ExternEngineInterface>,
+}
+
+impl BoxHandle for KernelScanDataIterator {}
+
+impl Drop for KernelScanDataIterator {
+    fn drop(&mut self) {
+        debug!("dropping KernelScanDataIterator");
+    }
+}
+
+/// Get an iterator over the data needed to perform a scan. This will return a
+/// [`KernelScanDataIterator`] which can be passed to [`kernel_scan_data_next`] to get the actual
+/// data in the iterator.
+///
+/// # Safety
+///
+/// Engine is responsible for passing a valid [`ExternEngineInterfaceHandle`] and [`Scan`]
+#[no_mangle]
+pub unsafe extern "C" fn kernel_scan_data_init(
+    engine_interface: *const ExternEngineInterfaceHandle,
+    scan: *mut Scan,
+) -> ExternResult<*mut KernelScanDataIterator> {
+    kernel_scan_data_init_impl(engine_interface, scan).into_extern_result(engine_interface)
+}
+
+unsafe fn kernel_scan_data_init_impl(
+    engine_interface: *const ExternEngineInterfaceHandle,
+    scan: *mut Scan,
+) -> DeltaResult<*mut KernelScanDataIterator> {
+    let engine_interface = unsafe { ArcHandle::clone_as_arc(engine_interface) };
+    let boxed_scan = unsafe { Box::from_raw(scan) };
+    let scan = boxed_scan.kernel_scan;
+    let scan_data = scan.scan_data(engine_interface.table_client().as_ref())?;
+    let data = KernelScanDataIterator {
+        data: Box::new(scan_data),
+        engine_interface,
+    };
+    Ok(data.into_handle())
+}
+
+/// # Safety
+///
+/// The iterator must be valid (returned by [kernel_scan_data_init]) and not yet freed by
+/// [kernel_scan_data_free]. The visitor function pointer must be non-null.
+#[no_mangle]
+pub unsafe extern "C" fn kernel_scan_data_next(
+    data: &mut KernelScanDataIterator,
+    engine_context: *mut c_void,
+    engine_visitor: extern "C" fn(
+        engine_context: *mut c_void,
+        engine_data: *mut EngineDataHandle,
+        selection_vector: KernelBoolSlice,
+    ),
+) -> ExternResult<bool> {
```

Comment:
hrmm. so the semantics are "rust like", meaning you first need to check the `tag` field. if that's `Ok_bool`, then the call succeeded, and the bool inside can be examined via looking at `ok`, which should be `false` only when the scan is done. So like:
```c

  // iterate scan files
  for (;;) {
    ExternResult_bool ok_res = kernel_scan_data_next(data_iter, &context, visit_data);
    if (ok_res.tag != Ok_bool) {
      printf("Failed to iterate scan data\n");
      return -1;
    } else if (!ok_res.ok) {
      break;
    }
  }
```

Reply:
I think I'm doing that https://github.com/samansmink/delta-kernel-testing/blob/e0b52161e36102615ef582cc7cf85fa8cce3c2de/src/functions/deltatable_scan.cpp#L118. I can try to make a reproduction if it helps

Reply:
repro would be cool. note it's possible that no data needs to be visited, because everything in that batch was removed. in which case, if you call back into kernel, you'll get no calls for actual data to read. we can probably optimize this and not bother returning it, and just try and advance the iterator again when we notice that there's nothing to actually scan in a batch.